doc/parameter_server.md

```diff
@@ -0,0 +1,98 @@
+# Parameter Server Design
+
+## Overview
+
+
+A typical parameter server architecture contains three roles:
+
+- master, responsible for creating/deleting/scheduling pservers and workers
+- pserver, responsible for provide parameter pull/push/optimize/checkpoint service
+- worker, responsible for training minibatches of data
+
+
+Please refer to:
+
+![parameter_server](./images/parameter_server.png)
+
+## Key concepts
+
+
+### Parameter sharding
+
+We usually shard big parameters of a neural network model to some nodes under a rule, in order to achieve network load banlance. There are several kinds of parameter to be handled seperately:
+
+- Very big embedding table: Embedding table is a collection of <item id, embedding vector> pairs. There is also a mapping from item id to pserver id, so the corresonding embedding vector will be stored at the certain pserver
+
+- Big dense tensor: If the dense tensor parameter exceeds certain size, it will be sliced into several subtensors. The parameter name combines subtensor id will become a unique key, and the value is the subtensor(Please note that the slice tensor operation is zero-copy)
+
+- Small dense tensor: The small dense tensor parameter will be stored at certain pserver wholely
+
```

Comment:
To make a complete discussion about model sharding, we also need to talk about the local model.

Indeed, we assume the local model is small enough that workers don't need to shard them. We need to explain why.

> The local model on workers and the global model on the parameter server(s) have the same size. In case that large embedding tables make the model size too big to fit in a single process's memory space, an intuitive solution is to save the model copies on an external storage service like Redis or Memcached.
>
> However, such a solution creates a complex workflow which introduces too many cross-node communications.
>
> We propose an alternative solution that doesn't rely on Redis or Memcachd.  Instead, the global model, including the large embedding tables, is sharded across parameter servers, and the local model contains only part of the embedding table -- the small fraction that is required to compute recent minibatch(es).

Reply:
Done.